src/lib.rs

```diff
@@ -737,31 +745,60 @@ impl<R: gimli::Reader> ParsedDwarf<R> {
             i.max_end = max;
         }
 
-        Ok(ParsedDwarf {
-            units: res_units,
-            unit_ranges,
-            sup_units: None,
-        })
+        Ok((unit_ranges, res_units))
     }
 
+    fn parse_sup(sections: &gimli::Dwarf<R>) -> Result<Vec<SupUnit<R>>, Error> {
+        let mut sup_units = Vec::new();
+        let mut units = sections.units();
+        while let Some(header) = units.next()? {
+            let offset = match header.offset().as_debug_info_offset() {
+                Some(offset) => offset,
+                None => continue,
+            };
+            let dw_unit = match sections.unit(header) {
+                Ok(dw_unit) => dw_unit,
+                Err(_) => continue,
+            };
+            sup_units.push(SupUnit { dw_unit, offset });
+        }
+        Ok(sup_units)
+    }
+
+    // Find the unit containing the given offset, and convert the offset into a unit offset.
     fn find_unit(
         &self,
         offset: gimli::DebugInfoOffset<R::Offset>,
         file: DebugFile,
-    ) -> Result<&ResUnit<R>, Error> {
-        let units = match file {
-            DebugFile::Primary => &self.units,
-            DebugFile::Supplementary => self
-                .sup_units
-                .as_ref()
-                .ok_or(gimli::Error::NoEntryAtGivenOffset)?,
+    ) -> Result<(&gimli::Unit<R>, gimli::UnitOffset<R::Offset>), Error> {
+        let unit = match file {
+            DebugFile::Primary => {
+                match self
+                    .units
+                    .binary_search_by_key(&offset.0, |unit| unit.offset.0)
```

Comment:
Duplicating the binary search logic strikes me as less clear than what was there before.

Reply:
True, but they are different types now so I didn't see how we could reuse it anymore.